Logical Model

wangcj05 opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Close #1202

##### What are the significant changes in functionality due to this change request?

Added LOGICAL MODEL CLASS
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

moosebuild commented:
Job [Test Fedora 28](https://civet.inl.gov/job/522857/) on 51ef98b : invalidated by @wangcj05

python environment issue

wangcj05 commented:
@mandd @alfoa Please take a look at this PR regarding the LogicalModel Implementation.

alfoa commented:
Can you update issue #1225? I know it is tedious but we did not implement an automatic system yet.

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/524181/) on f4b8567 : invalidated by @wangcj05

fetch issue

wangcj05 commented:
> Can you update issue #1225? I know it is tedious but we did not implement an automatic system yet.

Updated.

alfoa commented on tests/framework/hybridModel/test_logical_code.xml:
```diff
@@ -0,0 +1,105 @@
+    <description>
+      Example of usage of the Logical Model capability in RAVEN
+      This test is intended to check the functionality of Logical Model
+      with the generic code system.
```
can you be more descriptive here? What is the control function? and the models? sorry i know it is tedious but it can save us time later for updating the tests description

wangcj05 replied:
updated

alfoa commented on tests/framework/hybridModel/test_logical_model.xml:
```diff
@@ -0,0 +1,108 @@
+    <description>
+       Example of usage of the Logical Model capability in RAVEN
+       This test is intended to check the functionality of Logical Model
+       with multiple external models.
```
as above

wangcj05 replied:
updated

alfoa commented on tests/framework/hybridModel/test_logical_rom.xml:
```diff
@@ -0,0 +1,132 @@
+    <description>
+      Example of usage of the Logical Model capability in RAVEN
+      This test is intended to check the functionality of Logical Model
+      with ROMs.
```
same

wangcj05 replied:
updated

alfoa commented:
Sorry some "comments" are repeated but are just "place holders:

wangcj05 commented on doc/user_manual/logical_model.tex:
```diff
@@ -0,0 +1,177 @@
+
+\subsection{LogicalModel}
+\label{subsec:models_LogicalModel}
+The \textbf{LogicalModel} is a new \textit{Model} entity. This new Model is able to execute ROMs,
```
updated

wangcj05 commented on doc/user_manual/logical_model.tex:
```diff
@@ -0,0 +1,177 @@
+  executed.
+\end{itemize}
+
+\textbf{Example (ExternalModel):}
```
updated

wangcj05 commented on doc/user_manual/logical_model.tex:
```diff
@@ -0,0 +1,177 @@
+  return model
+\end{lstlisting}
+
+\textbf{Example (Code):}
```
updated

wangcj05 commented on doc/user_manual/logical_model.tex:
```diff
@@ -0,0 +1,177 @@
+  <Steps>
+    <MultiRun name="logicalModelCode">
+      <Input class="Files" type="">gen.one</Input>
+      <Input class="Files" type="">gen.two</Input>
```
Are you fine if I work on this issue in another pull request? Since the changes will also affect EnsembleModel and CodeInterface Entities. It may take longer and further discussion on the solution for this issue.

alfoa replied:
Yes. Please, reference a new issue about this.
Thank you.

wangcj05 replied:
Issue #1240 is created.

wangcj05 commented on framework/Models/HybridModels/LogicalModel.py:
```diff
@@ -0,0 +1,231 @@
+    """
+    HybridModelBase.__init__(self,runInfoDict)
+    self.printTag              = 'LogicalModel MODEL' # print tag
+    self.controlFunction       = None
```
updated

wangcj05 commented on framework/Models/HybridModels/LogicalModel.py:
```diff
@@ -0,0 +1,231 @@
+    HybridModelBase.localInputAndChecks(self, xmlNode)
+    paramInput = self.getInputSpecification()()
+    paramInput.parseNode(xmlNode)
+    for child in paramInput.subparts:
```
updated

wangcj05 commented on framework/Models/HybridModels/LogicalModel.py:
```diff
@@ -0,0 +1,231 @@
+    if self.modelInstances[modelToRun].type == 'Code':
+      codeInput = []
+      for elem in myInput:
+        if isinstance(elem, Files.File):
```
copied from previous comment: 
```
Are you fine if I work on this issue in another pull request? Since the changes will also affect EnsembleModel and CodeInterface Entities. It may take longer and further discussion on the solution for this issue.
```

wangcj05 commented on framework/Models/HybridModels/LogicalModel.py:
```diff
@@ -0,0 +1,231 @@
+    inputKwargs['prefix'] = modelToRun + utils.returnIdSeparator() + identifier
+    inputKwargs['uniqueHandler'] = self.name + identifier
+    moveOn = False
+    while not moveOn:
```
As we discussed during today's virtual meeting. Currently, this is probably the right way to handle additional client submission. We may need to extend the JobHandler capabilities in the future to make it more intelligent to manage jobs.

wangcj05 commented on framework/Models/HybridModels/LogicalModel.py:
```diff
@@ -0,0 +1,231 @@
+    inputKwargs['uniqueHandler'] = self.name + identifier
+    moveOn = False
+    while not moveOn:
+      if jobHandler.availability() > 0:
```
Same as the previous comments

_As we discussed during today's virtual meeting. Currently, this is probably the right way to handle additional client submission. We may need to extend the JobHandler capabilities in the future to make it more intelligent to manage jobs._

wangcj05 commented on framework/Models/HybridModels/LogicalModel.py:
```diff
@@ -0,0 +1,231 @@
+      else:
+        time.sleep(self.sleepTime)
+    while not jobHandler.isThisJobFinished(inputKwargs['prefix']):
+      time.sleep(self.sleepTime)
```
Same as the previous comments

_As we discussed during today's virtual meeting. Currently, this is probably the right way to handle additional client submission. We may need to extend the JobHandler capabilities in the future to make it more intelligent to manage jobs._

wangcj05 commented on framework/Models/HybridModels/LogicalModel.py:
```diff
@@ -0,0 +1,231 @@
+    finishedRun = jobHandler.getFinished(jobIdentifier=inputKwargs['prefix'], uniqueHandler=inputKwargs['uniqueHandler'])
+    evaluation = finishedRun[0].getEvaluation()
+    if isinstance(evaluation, rerror):
+      self.raiseAnError(RuntimeError, "The model", modelToRun, "identified by", finishedRun[0].identifier, "failed!")
```
I think we still need this. Since we submit the models through client, this check seems to me is more informative. It will indicate which sub-model is failed.

wangcj05 commented on tests/framework/hybridModel/logicalCode/control.py:
```diff
@@ -0,0 +1,28 @@
+
+def evaluate(self):
+  """
+    Method required by RAVEN to run this as an external model.
```
updated

wangcj05 commented on tests/framework/hybridModel/logicalCode/exp_code.py:
```diff
@@ -0,0 +1,63 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
```
This is needed, np.exp is used.

wangcj05 commented on tests/framework/hybridModel/logicalCode/exp_code.py:
```diff
@@ -0,0 +1,63 @@
+    @ In, inp, str, input file name
+    @ Out, None
+  """
+  inx = open(inp,'r')
```
updated as requested.

wangcj05 commented on tests/framework/hybridModel/logicalCode/exp_code.py:
```diff
@@ -0,0 +1,63 @@
+
+  dat = eval(x,y)
+
+  outf = open(case+'.csv','w')
```
updated as requested

wangcj05 commented on tests/framework/hybridModel/logicalCode/poly_code.py:
```diff
@@ -0,0 +1,63 @@
+  outf.writelines('step,i,j,x,y,poly\n')
+  for e in dat:
+    outf.writelines(','.join(str(i) for i in e)+'\n')
+  outf.close()
```
updated

wangcj05 commented on tests/framework/hybridModel/logicalRom/EM2linear.py:
```diff
@@ -0,0 +1,18 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
```
removed

wangcj05 commented on tests/framework/hybridModel/logicalRom/EM2linear.py:
```diff
@@ -0,0 +1,18 @@
+def run(self, Input):
+  # self.leftTemperature (boundary condition - left) self.rightTemperature (boundary condition - right)
+  self.averageTemperature = (self.leftTemperature + self.rightTemperature)/2.0
+  self.k = 38.23/(129.2 + self.averageTemperature) + 0.6077E-12*self.averageTemperature
```
this file is removed to avoid duplication

wangcj05 commented on tests/framework/hybridModel/logicalRom/control.py:
```diff
@@ -0,0 +1,30 @@
+
+def evaluate(self):
+  """
+    Method required by RAVEN to run this as an external model.
```
updated

wangcj05 commented on tests/framework/hybridModel/logicalRom/control.py:
```diff
@@ -0,0 +1,30 @@
+    Method required by RAVEN to run this as an external model.
+    @ In, self, object, object to store members on
+    @ Out, model, str, the name of external model that
+      will be executed by hybrid model
```
updated

wangcj05 commented on tests/framework/hybridModel/multiple/control.py:
```diff
@@ -0,0 +1,30 @@
+
+def evaluate(self):
+  """
+    Method required by RAVEN to run this as an external model.
```
updated

wangcj05 commented:
@alfoa I have addressed all your comments except how to handle files for codes. 
Are you fine if I work on this issue in another pull request? Since the changes will also affect EnsembleModel and CodeInterface Entities. It may take longer and further discussion on the solution for this issue.

alfoa commented:
when the test pass, we can merge this

wangcj05 commented:
A new issue is created to standardize the way to handle inputs for codes. #1240 @alfoa

moosebuild commented:
All [jobs](https://civet.inl.gov/event/55091/) on 17a8ae9 : invalidated by @wangcj05

fetch issue on Linux machines

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/531255/) on 17a8ae9 : invalidated by @wangcj05

fetch issue

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/535425/) on 0397d8b : invalidated by @joshua-cogliati-inl

restarted civet

wangcj05 commented:
@alfoa Could you merge this PR? All tests have passed now.

alfoa commented:
Checklist passed...

Merging...
